pkg/detectors/netsuite/netsuite.go

```diff
@@ -0,0 +1,227 @@
+
+	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+
+	if err != nil {
+		log.Fatalf("Error creating request: %v", err)
+		return false, err
+	}
+
+	// Set required headers
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Authorization", makeSignature("GET", url, "/services/rest/record/v1/metadata-catalog/customer", map[string]string{
+		"consumer_key":     consumerKey,
+		"consumer_secret":  consumerSecret,
+		"token_id":         tokenKey,
+		"token_secret":     tokenSecret,
+		"signature_method": "HMAC-SHA256",
+		"timestamp":        strconv.FormatInt(time.Now().Unix(), 10),
+		"nonce":            nonceGenerator.Nonce(),
+		"version":          "1.0",
+		"realm":            accountID,
+	}))
+
+	// Make the request
+	res, err := client.Do(req)
+	if err != nil {
+		return false, err
+	}
+	defer res.Body.Close()
+	switch res.StatusCode {
+	case http.StatusOK, http.StatusForbidden:
+		// 403 indicates lack of permission, but valid token (could be due to twitter free tier)
+		return true, nil
+	case http.StatusUnauthorized:
+		return false, nil
+	default:
+		return false, fmt.Errorf("unexpected HTTP response status %d", res.StatusCode)
+	}
+}
+
+func generateHMACSHA256(message, secret string) string {
+	// Create a new HMAC by defining the hash type and the key (as byte array)
+	h := hmac.New(sha256.New, []byte(secret))
+
+	// Write the message to it
+	h.Write([]byte(message))
+
+	// Get the resulting HMAC as a byte slice
+	hash := h.Sum(nil)
+
+	// Encode the byte slice to a hexadecimal string
+	return base64.StdEncoding.EncodeToString(hash)
+}
+
+func makeSignature(method, baseUrl, path string, params map[string]string) string {
+	baseString := method + "&" + url.QueryEscape(baseUrl+path) + "&" +
+		url.QueryEscape("oauth_consumer_key="+url.QueryEscape(params["consumer_key"])+
+			"&oauth_nonce="+url.QueryEscape(params["nonce"])+
+			"&oauth_signature_method="+url.QueryEscape(params["signature_method"])+
+			"&oauth_timestamp="+url.QueryEscape(params["timestamp"])+
+			"&oauth_token="+url.QueryEscape(params["token_id"])+
+			"&oauth_version="+url.QueryEscape(params["version"]))
+
+	key := url.QueryEscape(params["consumer_secret"]) + "&" + url.QueryEscape(params["token_secret"])
+
+	signature := generateHMACSHA256(baseString, key)
+
+	return "OAuth" + " " + "realm=\"" + params["realm"] + "\"," +
+		"oauth_consumer_key=\"" + params["consumer_key"] + "\"," +
+		"oauth_token=\"" + params["token_id"] + "\"," +
+		"oauth_signature_method=\"" + params["signature_method"] + "\"," +
+		"oauth_timestamp=\"" + params["timestamp"] + "\"," +
+		"oauth_nonce=\"" + params["nonce"] + "\"," +
+		"oauth_version=\"" + params["version"] + "\"," +
+		"oauth_signature=\"" + url.QueryEscape(signature) + "\""
+}
+
+// NetsuiteNoncer reads 11 bytes from crypto/rand and
+// returns those bytes as string
+type NetsuiteNoncer struct{}
+
+func (n NetsuiteNoncer) Nonce() string {
```

Comment:
question: How come you opted for a method on a struct instead of just a function? For example:
```
func nonce() (string, error) {
	const nonceSize = 11
	b := make([]byte, nonceSize)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}

	return base64.RawURLEncoding.EncodeToString(b)[:nonceSize], nil
}
```

Reply:
@ahrav thanks, its a good suggestion but I have not worked on code cleaning yet. Authentication on netsuite was a hassle due to signature, So I just copied the code from library to generate the nonce.